### PR TITLE
Merge clipDistance/cullDistance mask

### PR DESCRIPTION
**Pull request by JamesZhang-AMD**
Fixed a CTS failure KHR-GL46.clip_distance.functional:
When the ClipDistance is assigned to a negative value and the draw primitive type is point, the point can't be clipped, which requests to logical OR the clip mask into the cull mask, this change will tell the hardware to treat a clip mask export as both clip and cull mask so that make the point can be clipped.

### CODE REVIEW

File: lgc/patch/Gfx6ConfigBuilder.cpp [hunk 1]
```diff
@@ -492,8 +492,8 @@ template <typename T> void ConfigBuilder::buildVsRegConfig(ShaderStage shaderSta
 
     // Set fields CLIP_DIST_ENA_0 ~ CLIP_DIST_ENA_7 and CULL_DIST_ENA_0 ~ CULL_DIST_ENA_7
     unsigned paClVsOutCntl = GET_REG(&config->vsRegs, PA_CL_VS_OUT_CNTL);
-    paClVsOutCntl |= clipDistanceMask;
-    paClVsOutCntl |= (cullDistanceMask << 8);
+    unsigned finalMask = clipDistanceMask | cullDistanceMask;
```
Comment (amdrexu) [inline lgc/patch/Gfx6ConfigBuilder.cpp:495]:
```suggestion
    unsigned finalMask = clipDistanceMask | cullDistanceMask;
    // NOTE: <Explain why we should modify cullDistanceMask by taking clipDistanceMask into account>
    cullDistanceMask = clipDistanceMask | cullDistanceMask;
    finalClipCullMask = clipDistanceMask | (cullDistanceMask << 8);
```

File: lgc/patch/Gfx9ConfigBuilder.cpp [hunk 2]
```diff
@@ -2213,8 +2213,8 @@ template <typename T> void ConfigBuilder::setupPaSpecificRegisters(T *config) {
 
     // Set fields CLIP_DIST_ENA_0 ~ CLIP_DIST_ENA_7 and CULL_DIST_ENA_0 ~ CULL_DIST_ENA_7
     unsigned paClVsOutCntl = GET_REG(config, PA_CL_VS_OUT_CNTL);
-    paClVsOutCntl |= clipDistanceMask;
-    paClVsOutCntl |= (cullDistanceMask << 8);
+    unsigned finalMask = clipDistanceMask | cullDistanceMask;
```
Comment (amdrexu) [inline lgc/patch/Gfx9ConfigBuilder.cpp:2216]: The same as above

**Review by amdrexu — COMMENTED** [review]
Need to check the function RegisterMetadataBuilder::buildPaSpecificRegisters
 ```
    paClVsOutCntl[Util::Abi::PaClVsOutCntlMetadataKey::CullDistEna_0] = cullDistEna[0];
    paClVsOutCntl[Util::Abi::PaClVsOutCntlMetadataKey::CullDistEna_1] = cullDistEna[1];
    paClVsOutCntl[Util::Abi::PaClVsOutCntlMetadataKey::CullDistEna_2] = cullDistEna[2];
    paClVsOutCntl[Util::Abi::PaClVsOutCntlMetadataKey::CullDistEna_3] = cullDistEna[3];
    paClVsOutCntl[Util::Abi::PaClVsOutCntlMetadataKey::CullDistEna_4] = cullDistEna[4];
    paClVsOutCntl[Util::Abi::PaClVsOutCntlMetadataKey::CullDistEna_5] = cullDistEna[5];
    paClVsOutCntl[Util::Abi::PaClVsOutCntlMetadataKey::CullDistEna_6] = cullDistEna[6];
    paClVsOutCntl[Util::Abi::PaClVsOutCntlMetadataKey::CullDistEna_7] = cullDistEna[7];
```
GFX9+ will switch to this in the future.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 0f95b02cd7b7c4742734424d43b697a6391056ce
<details><summary>CTS tests (Failed: 0/213043)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40066/71017 (56.4%)</li>
<li>Failed: 0/71017 (0.0%)</li>
<li>Not Supported: 30951/71017 (43.6%)</li><li>Warnings: 0/71017 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40106/71007 (56.5%)</li>
<li>Failed: 0/71007 (0.0%)</li>
<li>Not Supported: 30901/71007 (43.5%)</li><li>Warnings: 0/71007 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Passed: 36212/71019 (51.0%)</li>
<li>Failed: 0/71019 (0.0%)</li>
<li>Not Supported: 34807/71019 (49.0%)</li><li>Warnings: 0/71019 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by tsymalla** [comment]
Can you please fix the commit message title? clipDisantce => clipDistance
Thanks!

File: lgc/patch/Gfx6ConfigBuilder.cpp (hunk 1 above)
Comment (JamesZhang-AMD) [reply lgc/patch/Gfx6ConfigBuilder.cpp:497]: Fixed.

File: lgc/patch/Gfx9ConfigBuilder.cpp (hunk 2 above)
Comment (JamesZhang-AMD) [reply lgc/patch/Gfx9ConfigBuilder.cpp:2218]: Fixed.

**Comment by JamesZhang-AMD** [comment]
> Need to check the function RegisterMetadataBuilder::buildPaSpecificRegisters
> 
> ```
>    paClVsOutCntl[Util::Abi::PaClVsOutCntlMetadataKey::CullDistEna_0] = cullDistEna[0];
>    paClVsOutCntl[Util::Abi::PaClVsOutCntlMetadataKey::CullDistEna_1] = cullDistEna[1];
>    paClVsOutCntl[Util::Abi::PaClVsOutCntlMetadataKey::CullDistEna_2] = cullDistEna[2];
>    paClVsOutCntl[Util::Abi::PaClVsOutCntlMetadataKey::CullDistEna_3] = cullDistEna[3];
>    paClVsOutCntl[Util::Abi::PaClVsOutCntlMetadataKey::CullDistEna_4] = cullDistEna[4];
>    paClVsOutCntl[Util::Abi::PaClVsOutCntlMetadataKey::CullDistEna_5] = cullDistEna[5];
>    paClVsOutCntl[Util::Abi::PaClVsOutCntlMetadataKey::CullDistEna_6] = cullDistEna[6];
>    paClVsOutCntl[Util::Abi::PaClVsOutCntlMetadataKey::CullDistEna_7] = cullDistEna[7];
> ```
> 
> GFX9+ will switch to this in the future.

Hi Rex,
I have double checked, that we need to apply the fix in RegisterMetadataBuilder::buildPaSpecificRegisters, thanks for pointing it.

**Comment by JamesZhang-AMD** [comment]
> Can you please fix the commit message title? clipDisantce => clipDistance Thanks!

Hi Thomas,
Thanks for pointing out the typo, I have fixed it.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 781414bd5a98aafbe0daf84ffe9b30a7e5257287
<details><summary>CTS tests (Failed: 0/213064)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40073/71024 (56.4%)</li>
<li>Failed: 0/71024 (0.0%)</li>
<li>Not Supported: 30951/71024 (43.6%)</li><li>Warnings: 0/71024 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40113/71014 (56.5%)</li>
<li>Failed: 0/71014 (0.0%)</li>
<li>Not Supported: 30901/71014 (43.5%)</li><li>Warnings: 0/71014 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Passed: 36219/71026 (51.0%)</li>
<li>Failed: 0/71026 (0.0%)</li>
<li>Not Supported: 34807/71026 (49.0%)</li><li>Warnings: 0/71026 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit cf19ebb53573cea4eba1ef39b45ab48edba55f6e
<details><summary>CTS tests (Failed: 0/142050)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40073/71024 (56.4%)</li>
<li>Failed: 0/71024 (0.0%)</li>
<li>Not Supported: 30951/71024 (43.6%)</li><li>Warnings: 0/71024 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Timeout or crash</li>

</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Passed: 36219/71026 (51.0%)</li>
<li>Failed: 0/71026 (0.0%)</li>
<li>Not Supported: 34807/71026 (49.0%)</li><li>Warnings: 0/71026 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 02503fe505914f27c4468b06cee5e91508c325c1
<details><summary>CTS tests (Failed: 0/142036)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40073/71010 (56.4%)</li>
<li>Failed: 0/71010 (0.0%)</li>
<li>Not Supported: 30937/71010 (43.6%)</li><li>Warnings: 0/71010 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Timeout or crash</li>

</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Passed: 36219/71026 (51.0%)</li>
<li>Failed: 0/71026 (0.0%)</li>
<li>Not Supported: 34807/71026 (49.0%)</li><li>Warnings: 0/71026 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 02503fe505914f27c4468b06cee5e91508c325c1
<details><summary>CTS tests (Failed: 374/213064)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39705/71024 (55.9%)</li>
<li><details><summary>Failed: 374/71024 (0.5%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.buffer.suballocation.storage_texel.index.create.binding_aliased
Stack trace: Flake

FAILURE: dEQP-VK.api.buffer.suballocation.transfer_src.transfer_dst.uniform_texel.storage_texel.uniform.storage.create.binding_aliased
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.blit_image.all_formats.generate_mipmaps.from_base_level.layercount_1.a8b8g8r8_unorm_pack32.optimal_optimal_linear
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.diff_layout_copy_before_resolving.8_bit_transfer_src_optimal_transfer_dst_optimal
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.whole_copy_before_resolving_transfer.2_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.core.depth_stencil_msaa_copy.whole.s8_uint_optimal_general_S_2_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.image_to_image.cube.cube_to_array_whole
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.image_to_image.dimensions.src6x16384_dst6x16384.r4g4b4a4_unorm_pack16.r4g4b4a4_unorm_pack16.optimal_optimal
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.resolve_image.diff_image_size.dst_256_256_11_2_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.core.clear_color_image.1d.optimal.remaining_array_layers.a4b4g4r4_unorm_pack16_clamp_input_multiple_subresourcerange
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.core.clear_color_image.2d.optimal.remaining_array_layers.r16g16_sfloat_64x11_multiple_subresourcerange
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.core.clear_color_image.2d.optimal.single_layer.a2r10g10b10_uint_pack32_200x180_multiple_subresourcerange
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.clear_color_image.1d.linear.multiple_layers.r8g8_unorm_200x1
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.clear_color_image.1d.optimal.single_layer.b4g4r4a4_unorm_pack16_71x1
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.clear_color_image.3d.optimal.single_layer.r8_unorm_clamp_input_multiple_subresourcerange
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.clear_depth_stencil_attachment.single_layer.d16_unorm_s8_uint_separate_layouts_depth_1x33
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.partial_clear_color_attachment.multiple_layers.r16g16_sfloat_64x11
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.partial_clear_depth_stencil_attachment.multiple_layers.d16_unorm_33x128
Stack trace: Flake

FAILURE: dEQP-VK.binding_model.buffer_device_address.set15.depth2.baseubo.convertchecku64.nostore.multi.std140.vert
Stack trace: Flake

FAILURE: dEQP-VK.binding_model.buffer_device_address.set3.depth1.baseubo.convertchecku64.nostore.multi.scalar.vert
Stack trace: Flake
...
```
</p></details></li>

<li>Not Supported: 30945/71024 (43.6%)</li><li>Warnings: 0/71024 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40113/71014 (56.5%)</li>
<li>Failed: 0/71014 (0.0%)</li>
<li>Not Supported: 30901/71014 (43.5%)</li><li>Warnings: 0/71014 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Passed: 36219/71026 (51.0%)</li>
<li>Failed: 0/71026 (0.0%)</li>
<li>Not Supported: 34807/71026 (49.0%)</li><li>Warnings: 0/71026 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 02503fe505914f27c4468b06cee5e91508c325c1
<details><summary>CTS tests (Failed: 0/142040)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Timeout or crash</li>

</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40113/71014 (56.5%)</li>
<li>Failed: 0/71014 (0.0%)</li>
<li>Not Supported: 30901/71014 (43.5%)</li><li>Warnings: 0/71014 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Passed: 36219/71026 (51.0%)</li>
<li>Failed: 0/71026 (0.0%)</li>
<li>Not Supported: 34807/71026 (49.0%)</li><li>Warnings: 0/71026 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 02503fe505914f27c4468b06cee5e91508c325c1
<details><summary>CTS tests (Failed: 0/213064)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40075/71024 (56.4%)</li>
<li>Failed: 0/71024 (0.0%)</li>
<li>Not Supported: 30949/71024 (43.6%)</li><li>Warnings: 0/71024 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40113/71014 (56.5%)</li>
<li>Failed: 0/71014 (0.0%)</li>
<li>Not Supported: 30901/71014 (43.5%)</li><li>Warnings: 0/71014 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Passed: 36219/71026 (51.0%)</li>
<li>Failed: 0/71026 (0.0%)</li>
<li>Not Supported: 34807/71026 (49.0%)</li><li>Warnings: 0/71026 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/patch/Gfx6ConfigBuilder.cpp (hunk 1 above)
Comment (nhaehnle) [reply lgc/patch/Gfx6ConfigBuilder.cpp:497]: Do you happen to know if there's a performance implication to setting more bits in this mask?

I find it difficult to understand the comment. How about:

> Note: Point primitives are only affected by the cull mask, so enable culling also based on clip distances.

**Review by nhaehnle — COMMENTED** [review]
Is there a CTS test or similar that is fixed by this? Please reference that in the commit message.

**Comment by JamesZhang-AMD** [comment]
> Is there a CTS test or similar that is fixed by this? Please reference that in the commit message.

Hi Nicolai,
This commit is to fix a CTS failure KHR-GL46.clip_distance.functional, I have added it in commit message.

File: lgc/patch/Gfx6ConfigBuilder.cpp (hunk 1 above)
Comment (JamesZhang-AMD) [reply lgc/patch/Gfx6ConfigBuilder.cpp:497]: Ok, I have updated to the new comments in the latest commit.

**Comment by amdvlk-admin** [comment]
### Test summary for commit ba0c1e127547408b6dfc760f0406099e29d58e75
<details><summary>CTS tests (Failed: 0/213047)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40074/71009 (56.4%)</li>
<li>Failed: 0/71009 (0.0%)</li>
<li>Not Supported: 30935/71009 (43.6%)</li><li>Warnings: 0/71009 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40112/71013 (56.5%)</li>
<li>Failed: 0/71013 (0.0%)</li>
<li>Not Supported: 30901/71013 (43.5%)</li><li>Warnings: 0/71013 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Passed: 36218/71025 (51.0%)</li>
<li>Failed: 0/71025 (0.0%)</li>
<li>Not Supported: 34807/71025 (49.0%)</li><li>Warnings: 0/71025 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**